pkg/runtime/codec.go

```diff
@@ -196,3 +196,79 @@ func (s base64Serializer) Decode(data []byte, defaults *unversioned.GroupVersion
 	}
 	return s.Serializer.Decode(out[:n], defaults, into)
 }
+
+// Assert that unversioned.GroupVersion and GroupVersions implement GroupVersioner
+var _ GroupVersioner = unversioned.GroupVersion{}
+var _ GroupVersioner = unversioned.GroupVersions{}
+var _ GroupVersioner = multiGroupVersioner{}
+
+type multiGroupVersioner struct {
+	target     unversioned.GroupVersion
+	groupKinds []unversioned.GroupKind
+}
+
+// NewMultiGroupVersioner creates a group versioner that returns gv for each specified group, and
+// prefers the first item in groups.
+func NewMultiGroupVersioner(gv unversioned.GroupVersion, groupKinds ...unversioned.GroupKind) GroupVersioner {
+	if len(groupKinds) == 0 || (len(groupKinds) == 1 && groupKinds[0].Group == gv.Group) {
+		return gv
+	}
+	return multiGroupVersioner{target: gv, groupKinds: groupKinds}
+}
+
+// VersionForGroupKind returns the target version if the provided src GroupKind matches any of the allowed
+// group kinds. If Kind is specified on src, only an equivalent Kind or empty kind will match.
+func (v multiGroupVersioner) VersionForGroupKind(src unversioned.GroupKind) (unversioned.GroupVersion, bool) {
+	// match any group kind with the same kind
+	if len(src.Kind) > 0 {
+		for _, kind := range v.groupKinds {
```

Comment:
Yeah, it looks like this encodes the faulty assumption that every kind exists in every groupversion...